Implementation for github triSYCL issue #126 and enterprise issue #4 and #7

yu810226 opened the pull request:
triSYCL issue #126: https://github.com/triSYCL/triSYCL/issues/126
enterprise issue #4: https://gitenterprise.xilinx.com/rkeryell/triSYCL/issues/4
enterprise issue #7: https://gitenterprise.xilinx.com/rkeryell/triSYCL/issues/7

yu810226 committed:
Modify LoopIdiomRecognize.cpp to skip loop idiom optimization for kernel and functions called in the kernel

This branch is aim to solve problem trigger in this example:
https://gitenterprise.xilinx.com/linyay/triSYCL/commit/2a0dbdb485293c32d1ce380a68d35ea897269b06#diff-4beb17f5a3dbac38d7cb72c001c1b27d

Since argument promotion can only be done with GEP instructions that
only used by loads. And if there are llvm.memcpy instruction, there
will always a bitcast instruction before to cast the pointer, so force
kernel and functions called in the kernel in per_kernel_caller.ll to
skip loop idiom opt pass.

yu810226 committed:
Modify SYCLArgsFlattening.cpp

This pass is modified for solving problem in example
https://github.com/yu810226/triSYCL/commit/70fb4606d71e79e74aecb8565e0fdbede886356b#diff-bbbcbcbc399a038ddcd05dbbdb685226

While the kernel is big enough, device compiler generated a function
inside instantiate_kernel function in pre_kernel.ll file leading to
argument flattening pass not working. This modification is aim to
solve this.

yu810226 committed:
Modify name and calling convention for SPIR functions

1. SYCLModifySPIRFuncName.cpp pass is added for modifying function
names for functions called in SYCL kernel.

2. In inSPIRation.cpp, add code to modify functions called in SYCL
kernel to have SPIR_FUNC calling convention.

yu810226 committed:
Rename basic block name to avoid $ produced

Solution for issue https://gitenterprise.xilinx.com/rkeryell/triSYCL/issues/7

keryell commented on lib/Transforms/Scalar/CMakeLists.txt:
```diff
@@ -54,6 +54,8 @@ add_llvm_library(LLVMScalarOpts
   StraightLineStrengthReduce.cpp
   StructurizeCFG.cpp
   TailRecursionElimination.cpp
+  ItaniumDemangle.cpp
```
I would keep all the SYCL-related passes in `lib/SYCL` instead of scattering them everywhere.

yu810226 replied:
Fixed. Thanks! :-)

keryell commented on lib/Transforms/Scalar/LoopIdiomRecognize.cpp:
```diff
@@ -219,6 +239,26 @@ static void deleteDeadInstruction(Instruction *I) {
+      }
+    }
+  }
+
```
I am lost about all the changes in this file.
Is it some old experimental code which has been replaced by https://github.com/keryell/llvm/pull/1?

yu810226 replied:
No, code added in this file is to force generated IR to be what argument promotion pass liked. Since argument promotion can only be done with GEP instructions that
only used by loads. And if there are llvm.memcpy instruction, there
will always a bitcast instruction before to cast the pointer. This makes argument promotion not working, so force kernels and functions called in the kernel in per_kernel_caller.ll to
skip loop idiom opt pass.

keryell commented on lib/Transforms/Scalar/SYCLKernel.cpp:
```diff
@@ -0,0 +1,112 @@
+}
+
+}
+}
```
Copy-paste code?

yu810226 replied:
Remove it now.

keryell commented on lib/Transforms/Scalar/ItaniumDemangle.cpp:
```diff
@@ -0,0 +1,4278 @@
+//===- ItaniumDemangle.cpp ------------------------------------------------===//
```
Copy-pasted file...

yu810226 replied:
Remove it now.

keryell commented on lib/SYCL/SYCLArgsFlattening.cpp:
```diff
@@ -332,6 +332,18 @@ PromoteArguments(CallGraphNode *CGN, CallGraph &CG,
     bool isSafeToPromote = (PtrArg->hasByValAttr() &&
        (isDenselyPacked(AgTy, DL) || !canPaddingBeAccessed(PtrArg)));
+    
+    for (Use &U : F->uses()) {
```
`auto &U` ?

keryell commented on lib/SYCL/SYCLArgsFlattening.cpp:
```diff
@@ -332,6 +332,18 @@ PromoteArguments(CallGraphNode *CGN, CallGraph &CG,
        (isDenselyPacked(AgTy, DL) || !canPaddingBeAccessed(PtrArg)));
+    
+    for (Use &U : F->uses()) {
+      CallSite CS(U.getUser());
```
C++11 `CS { ... }`

keryell commented on lib/SYCL/SYCLArgsFlattening.cpp:
```diff
@@ -332,6 +332,18 @@ PromoteArguments(CallGraphNode *CGN, CallGraph &CG,
+    
+    for (Use &U : F->uses()) {
+      CallSite CS(U.getUser());
+      if (CS.getInstruction() != nullptr) {
```
`if (auto I = CS.getInstruction())`

keryell commented on lib/SYCL/SYCLArgsFlattening.cpp:
```diff
@@ -332,6 +332,18 @@ PromoteArguments(CallGraphNode *CGN, CallGraph &CG,
+      CallSite CS(U.getUser());
+      if (CS.getInstruction() != nullptr) {
+        // If the function is called in kernel, force to do argument promotion
+        if (sycl::isKernel(*(CS.getInstruction()->getParent()->getParent()))) {
```
`if (sycl::isKernel(*(I->getParent()->getParent()))) {`

keryell commented on lib/SYCL/SYCLArgsFlattening.cpp:
```diff
@@ -332,6 +332,18 @@ PromoteArguments(CallGraphNode *CGN, CallGraph &CG,
+          DEBUG(dbgs() << "SYCL: " << F->getName() << "is called in kernel function.\n");
+	}
+      }
+    }
```
Probably not all the `}` are required.

keryell commented on lib/SYCL/SYCLArgsFlattening.cpp:
```diff
@@ -332,6 +332,18 @@ PromoteArguments(CallGraphNode *CGN, CallGraph &CG,
+        // If the function is called in kernel, force to do argument promotion
+        if (sycl::isKernel(*(CS.getInstruction()->getParent()->getParent()))) {
+          isSafeToPromote = true;
+          DEBUG(dbgs() << "SYCL: " << F->getName() << "is called in kernel function.\n");
```
We have to discus off-line because I to not understand the logic.
What happens if there are several nested calls?

keryell commented on lib/SYCL/SYCLArgsFlattening.cpp:
```diff
@@ -433,6 +446,10 @@ static bool PrefixIn(const IndicesVector &Indices,
     // prefix exists.
     //
     // This load is safe if any prefix of its operands is safe to load.
```
Now this comment is misplaced.

keryell commented on lib/SYCL/SYCLArgsFlattening.cpp:
```diff
@@ -433,6 +446,10 @@ static bool PrefixIn(const IndicesVector &Indices,
+    for (auto itr = Indices.begin(); itr != Indices.end(); itr++)
+      DEBUG(dbgs() << "PrefixIn Indices: " << *itr << "\n");
+    for (auto itr = Low->begin(); itr != Low->end(); itr++)
+      DEBUG(dbgs() << "PrefixIn Low: " << *itr << "\n");
```
There should be only 1  `DEBUG()` with the 2 loops inside.
Furthermore, why you do not use range-based for loop?

keryell commented on lib/SYCL/SYCLArgsFlattening.cpp:
```diff
@@ -383,8 +395,9 @@ PromoteArguments(CallGraphNode *CGN, CallGraph &CG,
-  if (ArgsToPromote.empty() && ByValArgsToTransform.empty())
+  if (ArgsToPromote.empty() && ByValArgsToTransform.empty()) {
     return nullptr;
+  }
```
Useless `{ }`

keryell commented on lib/SYCL/SYCLArgsFlattening.cpp:
```diff
@@ -513,6 +530,23 @@ static bool isSafeToPromoteArgument(Argument *Arg, bool isByValOrInAlloca,
+      }
+    }
+  }
+
```
All this code is *déjà vu*.
Create a function to use it twice?

keryell commented on lib/SYCL/SYCLArgsFlattening.cpp:
```diff
@@ -528,12 +562,14 @@ static bool isSafeToPromoteArgument(Argument *Arg, bool isByValOrInAlloca,
           for (User::op_iterator II = GEP->idx_begin(), IE = GEP->idx_end();
                II != IE; ++II)
-            if (ConstantInt *CI = dyn_cast<ConstantInt>(*II))
+            if (ConstantInt *CI = dyn_cast<ConstantInt>(*II)) {
```
Useless `{ }`

keryell commented on lib/SYCL/SYCLArgsFlattening.cpp:
```diff
@@ -528,12 +562,14 @@ static bool isSafeToPromoteArgument(Argument *Arg, bool isByValOrInAlloca,
+            } else {
               // We found a non-constant GEP index for this argument? Bail out
               // right away, can't promote this argument at all.
+	      DEBUG(dbgs()  << "SYCL: " << Arg->getName() << " in " << F->getName() << " used in non-constant GEP index.\n");
```
Indentation error.

keryell commented on lib/SYCL/SYCLArgsFlattening.cpp:
```diff
@@ -566,36 +602,49 @@ static bool isSafeToPromoteArgument(Argument *Arg, bool isByValOrInAlloca,
         // TODO: This runs the above loop over and over again for dead GEPs
         // Couldn't we just do increment the UI iterator earlier and erase the
         // use?
+	DEBUG(dbgs() << "SYCL: Dead GEP" << *GEP << "\n");
```
Indentation error.

keryell commented on lib/SYCL/SYCLArgsFlattening.cpp:
```diff
@@ -566,36 +602,49 @@ static bool isSafeToPromoteArgument(Argument *Arg, bool isByValOrInAlloca,
-        i != e; ++i)
-        if (ConstantInt *C = dyn_cast<ConstantInt>(*i))
+        i != e; ++i) {
+	DEBUG(dbgs() << "SYCL: " << Arg->getName() << " used in GEP: " << *GEP << "\n");
```
Indentation error.

keryell commented on lib/SYCL/SYCLArgsFlattening.cpp:
```diff
@@ -566,36 +602,49 @@ static bool isSafeToPromoteArgument(Argument *Arg, bool isByValOrInAlloca,
+        i != e; ++i) {
+	DEBUG(dbgs() << "SYCL: " << Arg->getName() << " used in GEP: " << *GEP << "\n");
+        if (ConstantInt *C = dyn_cast<ConstantInt>(*i)) {
+	  DEBUG(dbgs() << C->getSExtValue() << " constant extend value.\n");
```
Indentation error.

keryell commented on lib/SYCL/SYCLArgsFlattening.cpp:
```diff
@@ -566,36 +602,49 @@ static bool isSafeToPromoteArgument(Argument *Arg, bool isByValOrInAlloca,
           Operands.push_back(C->getSExtValue());
-        else
+        } else {
+	  DEBUG(dbgs() << "Not a constant operand GEP.\n");
```
Indentation error.

keryell commented on lib/SYCL/SYCLArgsFlattening.cpp:
```diff
@@ -566,36 +602,49 @@ static bool isSafeToPromoteArgument(Argument *Arg, bool isByValOrInAlloca,
         if (LoadInst *LI = dyn_cast<LoadInst>(GEPU)) {
           // Don't hack volatile/atomic loads
-          if (!LI->isSimple()) return false;
+          if (!LI->isSimple()) return false; 
```
Trailing space.

keryell commented on lib/SYCL/SYCLArgsFlattening.cpp:
```diff
@@ -566,36 +602,49 @@ static bool isSafeToPromoteArgument(Argument *Arg, bool isByValOrInAlloca,
         } else {
           // Other uses than load?
+	  DEBUG(dbgs() << "SYCL: " << Arg->getName() << " used in GEP: " << *GEP << " in " <<  F->getName() << "\n");
+	  DEBUG(dbgs() << "Has other uses than load.\n");
```
2 indentation errors.

keryell replied:
Furthermore, you can fuse the message output with only 1 `DEBUG()` everywhere in the following.

keryell commented on lib/SYCL/SYCLArgsFlattening.cpp:
```diff
@@ -630,16 +679,25 @@ static bool isSafeToPromoteArgument(Argument *Arg, bool isByValOrInAlloca,
 
     MemoryLocation Loc = MemoryLocation::get(Load);
-    if (AAR.canInstructionRangeModRef(BB->front(), *Load, Loc, MRI_Mod))
+    if (AAR.canInstructionRangeModRef(BB->front(), *Load, Loc, MRI_Mod) && !isRelatedToKernel) {
```
Add a comment to explain the change

keryell commented on lib/SYCL/SYCLArgsFlattening.cpp:
```diff
@@ -630,16 +679,25 @@ static bool isSafeToPromoteArgument(Argument *Arg, bool isByValOrInAlloca,
-      for (BasicBlock *TranspBB : inverse_depth_first_ext(P, TranspBlocks))
-        if (AAR.canBasicBlockModify(*TranspBB, Loc))
-          return false;
+      for (BasicBlock *TranspBB : inverse_depth_first_ext(P, TranspBlocks)) {
```
Useless `{`

keryell commented on lib/SYCL/SYCLArgsFlattening.cpp:
```diff
@@ -630,16 +679,25 @@ static bool isSafeToPromoteArgument(Argument *Arg, bool isByValOrInAlloca,
+          DEBUG(dbgs() << "SYCL: " << Arg->getName() << " used in "
+                << F->getName()
+                << " every path from the entry block to the load is not transparency.\n");
+    return false;
```
Indentation error.

keryell commented on lib/SYCL/SYCLArgsFlattening.cpp:
```diff
@@ -630,16 +679,25 @@ static bool isSafeToPromoteArgument(Argument *Arg, bool isByValOrInAlloca,
+        if (AAR.canBasicBlockModify(*TranspBB, Loc) && !isRelatedToKernel) {
+          DEBUG(dbgs() << "SYCL: " << Arg->getName() << " used in "
+                << F->getName()
+                << " every path from the entry block to the load is not transparency.\n");
```
All the left `<<` should be aligned together

keryell commented on lib/SYCL/SYCLArgsFlattening.cpp:
```diff
@@ -630,16 +679,25 @@ static bool isSafeToPromoteArgument(Argument *Arg, bool isByValOrInAlloca,
-        if (AAR.canBasicBlockModify(*TranspBB, Loc))
-          return false;
+      for (BasicBlock *TranspBB : inverse_depth_first_ext(P, TranspBlocks)) {
+        if (AAR.canBasicBlockModify(*TranspBB, Loc) && !isRelatedToKernel) {
```
Comment on the change?

keryell commented on lib/SYCL/SYCLModifySPIRFuncName.cpp:
```diff
@@ -0,0 +1,94 @@
+//
+//===----------------------------------------------------------------------===//
+//
+// Detect functions called in SYCL kernels and modify their names.  
```
How? Why?

keryell replied:
I do not understand why SPIR appears everywhere here, even in the file name.
Could be `SYCLCleanUpFuncName.cpp` instead?

keryell commented on lib/SYCL/SYCLModifySPIRFuncName.cpp:
```diff
@@ -0,0 +1,94 @@
+
+/// Switch on debug with set DebugFlag=0 or set DebugFlag=1 in debugger or with
+/// option -debug or -debug-only=SYCL-modify-SPIR-funcName
+#define DEBUG_TYPE "SYCL-modify-SPIR-func-name"
```
Does not match the name in comment

keryell commented on lib/SYCL/SYCLModifySPIRFuncName.cpp:
```diff
@@ -0,0 +1,94 @@
+        for (Use &U : F.uses()) {
+          CallSite CS(U.getUser());
+          if (CS.getInstruction() != nullptr) {
+            if (sycl::isKernel(*(CS.getInstruction()->getParent()->getParent()))) {
```
All this code is *déjà vu*.
Move this to a function in `lib/SYCL/SYCLKernel.cpp`?

keryell commented on lib/SYCL/SYCLModifySPIRFuncName.cpp:
```diff
@@ -0,0 +1,94 @@
+            if (sycl::isKernel(*(CS.getInstruction()->getParent()->getParent()))) {
+              DEBUG(dbgs() << F.getName() << "is called in kernel function. Force to change name.\n");
+              SYCLFuncCalledInKernelFound++;
+              F.setName("foo." + Twine(count++));  
```
Ahem... Naming the function `foo` is not very clear. :-)
Some more triSYCL-related name?
Also `Twine { ... }`

keryell commented on lib/SYCL/inSPIRation.cpp:
```diff
@@ -189,9 +201,30 @@ struct inSPIRation : public ModulePass {
+    F.setCallingConv(CallingConv::SPIR_FUNC);
+  }
+  
+  /// Rename basic block name to force not contain $ sign in the name
```
Rename basic block name to avoid $ sign in the name

keryell commented on lib/SYCL/inSPIRation.cpp:
```diff
@@ -189,9 +201,30 @@ struct inSPIRation : public ModulePass {
+  void renameBlockname(Function &F) {
+    int count = 0;
+    for (auto &B : F) {
+      B.setName("label_" + Twine(count++));
```
`Twine { ... }`

keryell commented on lib/SYCL/inSPIRation.cpp:
```diff
@@ -238,8 +271,24 @@ struct inSPIRation : public ModulePass {
       // \todo Put SPIR calling convention on declarations too
-      if (!F.isDeclaration() && sycl::isKernel(F))
-        kernelSPIRify(F);
+      if (!F.isDeclaration()) {
```
Useless `{`

keryell replied:
Otherwise, not clear why you changed the syntax...

yu810226 replied:
Since it causes a warning, I put the { for line 274. There is a chance to have ambiguous syntax for if.

keryell commented on lib/SYCL/inSPIRation.cpp:
```diff
@@ -238,8 +271,24 @@ struct inSPIRation : public ModulePass {
+      if (!F.isDeclaration()) {
+        if (sycl::isKernel(F)) {
+          kernelSPIRify(F);
+	  renameBlockname(F);
```
Indentation error.

keryell commented on lib/SYCL/inSPIRation.cpp:
```diff
@@ -238,8 +271,24 @@ struct inSPIRation : public ModulePass {
+	  renameBlockname(F);
+	} else {
+          // For function is called in SYCL kernels
+          // Put SPIR calling convention
```
Fix and unify this 2 line comment.

keryell commented on lib/SYCL/inSPIRation.cpp:
```diff
@@ -238,8 +271,24 @@ struct inSPIRation : public ModulePass {
+          for (Use &U : F.uses()) {
+            CallSite CS(U.getUser());
+            if (CS.getInstruction() != nullptr) {
+              if (sycl::isKernel(*(CS.getInstruction()->getParent()->getParent()))) {
```
*Déjà vu*

keryell commented on lib/SYCL/inSPIRation.cpp:
```diff
@@ -238,8 +271,24 @@ struct inSPIRation : public ModulePass {
+            if (CS.getInstruction() != nullptr) {
+              if (sycl::isKernel(*(CS.getInstruction()->getParent()->getParent()))) {
+                kernelCallFuncSPIRify(F);
+	        renameBlockname(F);
```
Indentation error.

keryell commented on lib/SYCL/inSPIRation.cpp:
```diff
@@ -238,8 +271,24 @@ struct inSPIRation : public ModulePass {
+	    }
+          }
+        }
+      }
```
I have the feeling that what we actually need is to do this on the transitive closure of all the functions called from a kernel.
So is it in the opposite direction?
And anyway, since after kernel code selection, there should be only functions that are called by some kernel or it would be intrinsics.
So I guess a lot of things could be simplified.

yu810226 replied:
Yes, what we want to do is to get transitive closure of functions called by kernel. And now it is doing in the opposite way: finding the call site which is from kernel.
And yes! After kernel code selection, there should be only functions called by kernels or intrinsics. It can be simplified to only test whether the function is intrinsic or not. Thanks! :-)

keryell reviewed (changes requested):
Great contribution.
But it looks like there should be different PR for each feature.
Several details are still to be improved.
Thank.

yu810226 committed:
Remove copy-paste files. Keep them under SYCL directory only.

yu810226 committed:
Simplified inSPIRation.cpp

yu810226 committed:
Minor fix for LoopIdiomRecognize.cpp

yu810226 committed:
Modify SYCL pass name and add new function in SYCLKernel.cpp

Modify pass SYCLModifySPIRFuncName.cpp to SYCLCleanUpFuncName.cpp
Change SYCL function called by kernel names to have sycl_func_ prefix.
Add new function detecting call site is kernel function in
SYCLKernel.cpp
Other minor fixes for comments or indentation.

yu810226 committed:
Fix SYCLArgsFlattening.cpp to deal with nested call

yu810226 committed:
Move code in SYCLCleanUpFuncName.cpp inside inSPIRation.cpp

yu810226 committed:
Modify function to test if function is called by kernel/ Minor fix for files

keryell commented on lib/SYCL/SYCLArgsFlattening.cpp:
```diff
@@ -63,11 +63,13 @@
 #include "llvm/Support/raw_ostream.h"
 #include <set>
+#include <vector>
+#include <algorithm>
```
All the `#include` should be alphabetically sorted.

keryell commented on lib/Transforms/Scalar/CMakeLists.txt:
```diff
@@ -54,6 +54,8 @@ add_llvm_library(LLVMScalarOpts
   StructurizeCFG.cpp
   TailRecursionElimination.cpp
+  ../../SYCL/ItaniumDemangle.cpp
+  ../../SYCL/SYCLKernel.cpp
```
It is not clear to me why you need this in Transforms/Scalar

keryell replied:
Ah, I realize that you made some change to some `Transforms/Scalar` passes...
Could it be possible to have only changes in `lib/SYCL` ?
What is the trade-off ?

yu810226 replied:
From your idea, try to execute loop idiom pass after SYCL argument flattening pass for the host. Thanks!

keryell commented on lib/SYCL/SYCLArgsFlattening.cpp:
```diff
@@ -332,12 +334,48 @@ PromoteArguments(CallGraphNode *CGN, CallGraph &CG,
+        }
+      }
+    }
+
```
I am curious about the complexity of this method based on callee-graph compared to a more direct approach with a single DFS on the call graph starting from kernels and just adding the function to a set of called-from-kernel functions that you can just use later when required. This could go to lib/SYCL/SYCLKernel.cpp

yu810226 replied:
Try to move the code in DFS algorithm and mark all functions that have kernels as ancestors. And in SYCL argument flattening pass, we can easily find these marked function, and force their arguments flattened.

keryell commented on lib/SYCL/SYCLArgsFlattening.cpp:
```diff
@@ -513,6 +556,49 @@ static bool isSafeToPromoteArgument(Argument *Arg, bool isByValOrInAlloca,
+      }
+    }
+  }
+
```
Déjà vu

keryell reviewed (changes requested):
Thanks for the cleaning!
There are still a few things to discuss.

yu810226 committed:
Remove code that force not executing loop idiom pass for some functions in host side

For host, skip loop idiom pass here, and execute it after SYCL
arguments flattening pass.

yu810226 committed:
Add new functions to support detecting functions that have kernel as an ancestor

yu810226 committed:
Modify SYCL Argument flattening pass

1. Modify SYCL Argument flattening pass for better complexity to find out functions having kernel as an ancestor.
2. Add comment for SYCL Argument flattening pass.
3. Clean up code for SYCLKernel.cpp

yu810226 commented:
This PR needs to wait for #1 and #2 finish merging.

keryell commented on lib/SYCL/inSPIRation.cpp:
```diff
@@ -235,11 +262,43 @@ struct inSPIRation : public ModulePass {
+        } else if (!F.isIntrinsic()) {
+          // After kernel code selection, there are only two kinds of functions
+          // left: funcions called by kernels or LLVM intrinsic functions.
+          // For functions called in SYCL kernels. Put SPIR calling convention.
```
kernels, put

keryell commented on lib/Transforms/IPO/PassManagerBuilder.cpp:
```diff
@@ -291,7 +292,8 @@ void PassManagerBuilder::addFunctionSimplificationPasses(
   addInstructionCombiningPass(MPM);
   MPM.add(createIndVarSimplifyPass());        // Canonicalize indvars
-  MPM.add(createLoopIdiomPass());             // Recognize idioms like memset.
+  if (EnableLoopIdiom)
```
All this code seems to have been committed already.
I guess there is some history rewriting to be done?

keryell commented on include/llvm/SYCL.h:
```diff
@@ -24,6 +27,22 @@ namespace sycl {
 
+/// Test if a function is in the list of having kernel ancestor
+bool isInHasKernelAncestorFunctionList(Function &F,
+                       std::vector<Function *> &hasKernelAncestorFunctionList);
```
Bad data structure, function name and argument name.

keryell commented on include/llvm/SYCL.h:
```diff
@@ -24,6 +27,22 @@ namespace sycl {
+
+/// Test if a function has ancestor kernel
+bool hasAncestorKernel(Function &F,
+                       std::vector<Function *> &hasKernelAncestorFunctionList);
```
Bad data structure  and argument name.

keryell commented on include/llvm/SYCL.h:
```diff
@@ -24,6 +27,22 @@ namespace sycl {
+
+/// Compute CallGraph to record all functions that have ancestor kernel
+void computeAncestorNode(CallGraphSCC &SCC, CallGraph &CG,
+                         std::vector<Function *> &hasKernelAncestorFunctionList);
```
Bad data structure  and argument name.

keryell commented on include/llvm/SYCL.h:
```diff
@@ -24,6 +27,22 @@ namespace sycl {
+
+/// Update functions that have ancestor kernel list when new CallGraphNode created in CallGraph
+void updateHasKernelAncestorFunctionList (CallGraphNode &NewNode,
+                                          std::vector<Function *> &hasKernelAncestorFunctionList);
```
Bad data structure  and argument name.

keryell commented on lib/SYCL/SYCLKernel.cpp:
```diff
@@ -71,6 +79,59 @@ bool isKernel(const Function &F) {
+/// Test if a function is in the list of having kernel ancestor
+bool isInHasKernelAncestorFunctionList(Function &F,
+                       std::vector<Function *> &hasKernelAncestorFunctionList) {
+  if (std::find(hasKernelAncestorFunctionList.begin(), hasKernelAncestorFunctionList.end(), &F)
```
What is the complexity ? Wrong data structure...
I guess what you really want is a set.

keryell replied:
What about `llvm/ADT/SmallPtrSet.h` ? It is already used a lot in ArgumentPromotion.cpp

keryell commented on lib/SYCL/SYCLKernel.cpp:
```diff
@@ -71,6 +79,59 @@ bool isKernel(const Function &F) {
+  return false;
+}
+
+/// Compute CallGraph to record all functions that have ancestor kernel
```
that have the kernel as ancestor ?

keryell commented on lib/SYCL/SYCLKernel.cpp:
```diff
@@ -71,6 +79,59 @@ bool isKernel(const Function &F) {
+/// Compute CallGraph to record all functions that have ancestor kernel
+void computeAncestorNode(CallGraphSCC &SCC, CallGraph &CG,
+                         std::vector<Function *> &hasKernelAncestorFunctionList) {
+  // Find the CGN the kernel function belongs to
```
What is a CGN?
Add a `.` at ends of sentence. Otherwise I cannot parse your comments. :-)

keryell commented on lib/SYCL/SYCLKernel.cpp:
```diff
@@ -71,6 +79,59 @@ bool isKernel(const Function &F) {
+void computeAncestorNode(CallGraphSCC &SCC, CallGraph &CG,
+                         std::vector<Function *> &hasKernelAncestorFunctionList) {
+  // Find the CGN the kernel function belongs to
+  // DFS algorithm start from the kernel function CGN to record all the function
```
starts
functions

keryell commented on lib/SYCL/SYCLKernel.cpp:
```diff
@@ -71,6 +79,59 @@ bool isKernel(const Function &F) {
+                         std::vector<Function *> &hasKernelAncestorFunctionList) {
+  // Find the CGN the kernel function belongs to
+  // DFS algorithm start from the kernel function CGN to record all the function
+  // has ancestor of it
```
which have the kernel as ancestor.
Or "which are callable from the kernel."?

keryell commented on lib/SYCL/SYCLKernel.cpp:
```diff
@@ -71,6 +79,59 @@ bool isKernel(const Function &F) {
+}
+
+/// Update functions that have ancestor kernel list when new CallGraphNode created in CallGraph
+void updateHasKernelAncestorFunctionList (CallGraphNode &NewNode,
```
Spurious space.

keryell commented on lib/SYCL/SYCLArgsFlattening.cpp:
```diff
@@ -620,6 +665,10 @@ static bool isSafeToPromoteArgument(Argument *Arg, bool isByValOrInAlloca,
   // the function to each of the load instructions.
 
+  // However, if the fuction has kernel as an ancestor, it is guaranteed to not
+  // be modified from entry of the load instructions. isRelatedToKernel is
```
"from entry of the load instructions"? What do you mean? Compare to the similar original comments.

keryell commented on lib/SYCL/SYCLArgsFlattening.cpp:
```diff
@@ -628,18 +677,25 @@ static bool isSafeToPromoteArgument(Argument *Arg, bool isByValOrInAlloca,
     // the load itself.
     BasicBlock *BB = Load->getParent();
-
+    
```
Spurious spacing.

keryell commented on lib/SYCL/inSPIRation.cpp:
```diff
@@ -235,11 +262,43 @@ struct inSPIRation : public ModulePass {
+      if (!F.isDeclaration()) {
+        if (sycl::isKernel(F)) {
+          kernelSPIRify(F);
+          
```
Spurious spacing. Please configure your editor correctly...

keryell commented on lib/SYCL/inSPIRation.cpp:
```diff
@@ -235,11 +262,43 @@ struct inSPIRation : public ModulePass {
+          // left: funcions called by kernels or LLVM intrinsic functions.
+          // For functions called in SYCL kernels. Put SPIR calling convention.
+          kernelCallFuncSPIRify(F);
+          
```
Spurious spacing. Please configure your editor correctly...

keryell commented on lib/Transforms/Scalar/LoopIdiomRecognize.cpp:
```diff
@@ -58,6 +58,7 @@
 #include "llvm/Transforms/Utils/BuildLibCalls.h"
 #include "llvm/Transforms/Utils/Local.h"
 #include "llvm/Transforms/Utils/LoopUtils.h"
+
```
While it is cleaner, we do not need to change LLVM for this...

keryell commented on lib/Transforms/Scalar/LoopIdiomRecognize.cpp:
```diff
@@ -219,6 +220,7 @@ static void deleteDeadInstruction(Instruction *I) {
 
 bool LoopIdiomRecognize::runOnLoop(Loop *L) {
   CurLoop = L;
+
```
While it is cleaner, we do not need to change LLVM for this...

keryell commented on lib/SYCL/SYCLKernel.cpp:
```diff
@@ -71,6 +79,59 @@ bool isKernel(const Function &F) {
+  // DFS algorithm start from the kernel function CGN to record all the function
+  // has ancestor of it
+  // for (CallGraphNode *I : SCC) {} will run in bottom-up order
+  for (auto SCCI = scc_begin(&CG); !SCCI.isAtEnd();
```
Is there a need for this loop with iterators?

keryell commented on lib/SYCL/SYCLKernel.cpp:
```diff
@@ -71,6 +79,59 @@ bool isKernel(const Function &F) {
+       ++SCCI) {
+    const std::vector<CallGraphNode*> &nextSCC = *SCCI;
+    for (auto I = nextSCC.begin(), E = nextSCC.end(); I != E; ++I) {
+      auto *F = (*I)->getFunction();
```
Should be in the following `if`

keryell commented on lib/SYCL/SYCLKernel.cpp:
```diff
@@ -71,6 +79,59 @@ bool isKernel(const Function &F) {
+  for (auto SCCI = scc_begin(&CG); !SCCI.isAtEnd();
+       ++SCCI) {
+    const std::vector<CallGraphNode*> &nextSCC = *SCCI;
+    for (auto I = nextSCC.begin(), E = nextSCC.end(); I != E; ++I) {
```
Is there a need for this loop with iterators?

keryell commented on lib/SYCL/SYCLKernel.cpp:
```diff
@@ -71,6 +79,59 @@ bool isKernel(const Function &F) {
+  // for (CallGraphNode *I : SCC) {} will run in bottom-up order
+  for (auto SCCI = scc_begin(&CG); !SCCI.isAtEnd();
+       ++SCCI) {
+    const std::vector<CallGraphNode*> &nextSCC = *SCCI;
```
Needed? Specially without `auto const &`?

keryell commented on lib/SYCL/inSPIRation.cpp:
```diff
@@ -235,11 +262,43 @@ struct inSPIRation : public ModulePass {
   /// Visit all the functions of the module
   bool runOnModule(Module &M) override {
+    // funcCount is for naming new name for each function called in kernel
+    int funcCount = 0;
```
I guess it would not work on some extern functions... We will fix that when we have the problem... :-)

keryell reviewed (changes requested):
Quite interesting.
There are still some room for cleaning and simplification.
Thanks.

yu810226 committed:
Minor fix for inSPIRation.cpp and LoopIdiomRecognize.cpp

Fix Spurious space and typo

yu810226 committed:
Fix for SYCL.h, SYCLArgsFlattening.cpp, and SYCLKernel.cpp

Change data structure from vector to SmallPtrSet.
Change function name, argument name and add comments for more clearer
explanation of the code.
fix Spurious space and typo

yu810226 committed:
Minor fix for lib/SYCL/inSPIRation.cpp

Since there would be a warning that the syntax might be ambiguous if we remove { here, put it back.

yu810226 committed:
Minor fix for SYCLKernel.cpp

Since we change the data structure from vector to SmallPtrSet. Change
push_back to insert.
Fix for ->

keryell commented on lib/SYCL/SYCLKernel.cpp:
```diff
@@ -7,19 +7,27 @@
 // Wait for LLVM 4.0...
 // #include "llvm/Demangle/Demangle.h"
 #include "llvm/IR/Function.h"
+#include "llvm/IR/CallSite.h"
```
Should be 1 line above

keryell commented on lib/SYCL/SYCLKernel.cpp:
```diff
@@ -71,6 +79,48 @@ bool isKernel(const Function &F) {
+    CallSite CS{U.getUser()};
+    if (auto I = CS.getInstruction()) {
+      auto parent = I->getParent()->getParent();
+      if (FunctionsCalledByKernel.count(parent))
```
I guess you can replace 4 lines by
```
return FunctionsCalledByKernel.count(parent);
```

keryell reviewed (changes requested):
Almost there!

yu810226 committed:
Minor fix for SYCLKernel.cpp

1. Reorder the include headers to follow the dictionary order.
2. Simplified the syntax.

keryell commented:
Great improvement!
Thanks.
